editor/src/viewport_tools/tool.rs

```diff
@@ -31,8 +31,89 @@ pub struct DocumentToolData {
 	pub secondary_color: Color,
 }
 
-pub trait ToolCommon: for<'a> MessageHandler<ToolMessage, ToolActionHandlerData<'a>> + PropertyHolder {}
-impl<T> ToolCommon for T where T: for<'a> MessageHandler<ToolMessage, ToolActionHandlerData<'a>> + PropertyHolder {}
+#[derive(Clone, Debug)]
+pub struct SignalToMessageMap {
+	pub document_dirty: Option<ToolMessage>,
+	pub selection_changed: Option<ToolMessage>,
+	pub tool_abort: Option<ToolMessage>,
+}
+
+pub trait ToolTransition {
+	fn signal_to_message_map(&self) -> SignalToMessageMap;
+	fn activate(&self, responses: &mut VecDeque<Message>) {
+		let shared_messages = self.signal_to_message_map();
```

Comment:
I honestly feel like it is less of a special case. When writing tools I think it feels far less "magical" than it did before. Instead of having to map these messages in a different file, almost everything regarding a tool is in one location. In addition, it removes some of the complex piping previously required in `tool_message_handler.rs`

Reply:
Sure was just a gut feeling